File: gpio4j/src/main/java/org/overworld/gpio4j/serial/wire/Notify.java
package org.overworld.gpio4j.serial.wire;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import org.overworld.gpio4j.serial.LineByte;
import org.overworld.gpio4j.serial.MessageType;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.With;

@NoArgsConstructor
@AllArgsConstructor
@With
@Data
public class Notify implements Message {

	private Integer pin;
	private Message data;
	private Long timestamp;

	@Override
	public Notify parse(ByteArrayInputStream wireData) throws IOException {
		
		// if (wireData.length != 2) throw new IllegalStateException("Incorrect number of bytes");
		// TODO Don't know length, need better check
		
		pin = (new LineByte(wireData.readNBytes(2)).asInt());
		
		data = MessageType.forId(wireData.read()).get().newInstance();
		data.parse(wireData);
		
		long l = (new LineByte(wireData.readNBytes(2))).asInt();
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();		
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();		
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();		
		l = (l << 8) | (new LineByte(wireData.readNBytes(2))).asInt();
		timestamp = l;

		return this;
	}

	@Override
	public ByteArrayOutputStream wireData() {
		
		if (pin == null) throw new IllegalStateException("Pin is not set when generating wire data");
		if (data == null) throw new IllegalStateException("Data is not set when generating wire data");
		if (timestamp == null) throw new IllegalStateException("Timestamp is not set when generating wire data");
		
		ByteArrayOutputStream result = new ByteArrayOutputStream();
		result.writeBytes((new LineByte(pin)).getHiLo());
		result.write(data.getType().getId());
		result.writeBytes(data.wireData().toByteArray());

		result.writeBytes(new LineByte((byte)((timestamp & 0xFF00000000000000L) >>> 56)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x00FF000000000000L) >>> 48)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x0000FF0000000000L) >>> 40)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x000000FF00000000L) >>> 32)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x00000000FF000000L) >>> 24)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x0000000000FF0000L) >>> 16)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x000000000000FF00L) >>> 8)).getHiLo());
		result.writeBytes(new LineByte((byte)((timestamp & 0x00000000000000FFL))).getHiLo());
		return result;
	}

	@Override
	public MessageType getType() {
		return MessageType.NOTIFY;
	}
}

File: gpioBoardDriver/serial/Notify.cpp
#include "Notify.h"

class Notify : public Message {
private:
    int pin;
    std::unique_ptr<Message> data;
    int64_t timestamp;

public:
    Notify() = default;
    Notify(int pin, std::unique_ptr<Message> data, int64_t timestamp)
        : pin(pin), data(std::move(data)), timestamp(timestamp) {}

    Notify& parse(std::istream& wireData) override {
        std::vector<uint8_t> buffer(2);
        wireData.read(reinterpret_cast<char*>(buffer.data()), 2);
        pin = LineByte(buffer).asInt();

        int messageTypeId = wireData.get();
        auto messageType = MessageType::forId(messageTypeId);
        if (!messageType) {
            throw std::runtime_error("Invalid message type");
        }
        data = messageType->newInstance();
        data->parse(wireData);

        int64_t l = 0;
        for (int i = 0; i < 8; ++i) {
            wireData.read(reinterpret_cast<char*>(buffer.data()), 2);
            l = (l << 8) | LineByte(buffer).asInt();
        }
        timestamp = l;

        return *this;
    }

    std::vector<uint8_t> wireData() const override {
        if (pin == 0) throw std::runtime_error("Pin is not set when generating wire data");
        if (!data) throw std::runtime_error("Data is not set when generating wire data");
        if (timestamp == 0) throw std::runtime_error("Timestamp is not set when generating wire data");

        std::vector<uint8_t> result;
        auto pinBytes = LineByte(pin).getHiLo();
        result.insert(result.end(), pinBytes.begin(), pinBytes.end());

        result.push_back(static_cast<uint8_t>(data->getType().getId()));

        auto dataBytes = data->wireData();
        result.insert(result.end(), dataBytes.begin(), dataBytes.end());

        for (int i = 7; i >= 0; --i) {
            auto byteValue = static_cast<uint8_t>((timestamp >> (i * 8)) & 0xFF);
            auto timestampBytes = LineByte(byteValue).getHiLo();
            result.insert(result.end(), timestampBytes.begin(), timestampBytes.end());
        }

        return result;
    }

    MessageType getType() const override {
        return MessageType::NOTIFY;
    }

    // Getters and setters
    int getPin() const { return pin; }
    void setPin(int p) { pin = p; }

    const Message* getData() const { return data.get(); }
    void setData(std::unique_ptr<Message> d) { data = std::move(d); }

    int64_t getTimestamp() const { return timestamp; }
    void setTimestamp(int64_t t) { timestamp = t; }
};
